vcr/holder/wallet.go

```diff
@@ -69,6 +74,65 @@ func New(
 	}
 }
 
+// BuildParams contains the parameters that will be included in the signature of the verifiable presentation
+type BuildParams struct {
+	Audience string
+	Expires  time.Time
+	Nonce    string
+}
+
+func (h wallet) BuildSubmission(ctx context.Context, walletDID did.DID, presentationDefinition pe.PresentationDefinition, acceptedFormats map[string]map[string][]string, params BuildParams) (*vc.VerifiablePresentation, *pe.PresentationSubmission, error) {
+	// get VCs from own wallet
+	credentials, err := h.List(ctx, walletDID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve wallet credentials: %w", err)
+	}
+
+	// match against the wallet's credentials
+	// if there's a match, create a VP and call the token endpoint
+	// If the token endpoint succeeds, return the access token
+	// If no presentation definition matches, return a 412 "no matching credentials" error
+	builder := presentationDefinition.PresentationSubmissionBuilder()
+	builder.AddWallet(walletDID, credentials)
+
+	// Find supported VP format, matching support from:
+	// - what the local Nuts node supports
+	// - the presentation definition "claimed format designation" (optional)
+	// - the verifier's metadata (optional)
+	formatCandidates := credential.OpenIDSupportedFormats(oauth.DefaultOpenIDSupportedFormats())
+	formatCandidates = formatCandidates.Match(credential.OpenIDSupportedFormats(acceptedFormats))
+	if presentationDefinition.Format != nil {
+		formatCandidates = formatCandidates.Match(credential.DIFClaimFormats(*presentationDefinition.Format))
+	}
+	format := pe.ChooseVPFormat(formatCandidates.Map)
+	if format == "" {
+		return nil, nil, errors.New("requester, verifier (authorization server metadata) and presentation definition don't share a supported VP format")
+	}
+	presentationSubmission, signInstructions, err := builder.Build(format)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to build presentation submission: %w", err)
+	}
+	if signInstructions.Empty() {
+		return nil, nil, ErrNoCredentials
+	}
+
+	// todo: support multiple wallets
+	vp, err := h.BuildPresentation(ctx, signInstructions[0].VerifiableCredentials, PresentationOptions{
+		Format: format,
+		ProofOptions: proof.ProofOptions{
+			Created:   time.Now(),
+			Challenge: &params.Nonce,
+			Domain:    &params.Audience,
+			Expires:   &params.Expires,
+			Nonce:     &params.Nonce,
```

Comment:
Add `params.Challenge` so you can set whatever is appropriate? Adding params.Nonce to both sounds like something that could break with other implementations.

Reply:
I'd rather keep it the same for now and reevaluate with the updated specs for proofs later this year